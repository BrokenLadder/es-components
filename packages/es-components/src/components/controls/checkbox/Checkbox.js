import React from 'react';
import PropTypes from 'prop-types';
import styled, { withTheme } from 'styled-components';

import CheckboxLabel from '../../containers/checkboxLabel/CheckboxLabel';

const CheckboxInput = styled.input`
  clip: rect(0, 0, 0, 0);
  pointer-events: none;
  position: absolute;

  &:focus ~ .es-checkbox__fill {
    box-shadow: 0 0 3px 3px ${props => props.theme.colors.inputFocus};
    &:after {
      border-color: ${({ checked, theme }) =>
        checked ? theme.colors.white : theme.colors.gray3};
    }
  }
`;

const CheckboxDisplay = styled.span`
  background: ${props => props.theme.colors.white};
  border: 3px solid ${props => props.theme.colors.gray8};
  border-radius: 4px;
  box-sizing: border-box;
  cursor: pointer;
  height: 25px;
  left: 10px;
  position: absolute;
  top: 0.55em;
  transition: all 0.25s linear;
  width: 25px;

  @media (min-width: ${props => props.theme.screenSize.tablet}) {
    left: 0;
    top: 0.35em;
  }

  &:after {
    background: transparent;
    border: 3px solid ${props => props.theme.colors.white};
    border-right: none;
    border-top: none;
    box-sizing: border-box;
    content: '';
    display: block;
    height: 9px;
    margin: 3px 0 0 2px;
    transform: rotate(-45deg);
    transition: border 0.25s linear;
    width: 15px;
  }
`;

const AdditionalHelpContent = styled.div`
  font-size: ${props => props.theme.sizes.baseFontSize};
  font-weight: 400;
  margin: 10px 0 10px 0;
  position: relative;
  right: 32px;
`;

function Checkbox({
  name,
  labelText,
  validationState,
  additionalHelpContent,
  theme,
  ...checkboxProps
}) {
  const helpId = additionalHelpContent ? `${name}-help` : undefined;
  const additionalHelp = additionalHelpContent && (
    <AdditionalHelpContent
      id={helpId}
      className="es-checkbox__help"
      validationState={validationState}
    >
      {additionalHelpContent}
    </AdditionalHelpContent>
  );

  return (
    <CheckboxLabel
      className="es-checkbox"
      validationState={validationState}
      checked={checkboxProps.checked}
      disabled={checkboxProps.disabled}
    >
      <CheckboxInput
        name={name}
        type="checkbox"
        focusBorderColor={theme.colors.inputFocus}
        {...checkboxProps}
      />
      <CheckboxDisplay className="es-checkbox__fill" />
      {labelText}
      {additionalHelp}
    </CheckboxLabel>
  );
}

Checkbox.propTypes = {
  /** The name of the checkbox */
  name: PropTypes.string.isRequired,
  labelText: PropTypes.string.isRequired,
  /** Display checkbox with contextual state colorings */
  validationState: PropTypes.oneOf(['default', 'success', 'warning', 'danger']),
  /** Content to display underneath the check box */
  additionalHelpContent: PropTypes.node,
  /**
   * Theme object used by the ThemeProvider,
   * automatically passed by any parent component using a ThemeProvider
   */
  theme: PropTypes.object.isRequired
};

Checkbox.defaultProps = {
  validationState: 'default',
  additionalHelpContent: undefined
};

export default withTheme(Checkbox);
