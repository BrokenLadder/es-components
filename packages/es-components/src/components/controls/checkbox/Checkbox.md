```
<div>
  <Checkbox name="uncheckedByDefault" labelText="I am unchecked by default" />
  <Checkbox name="checkedByDefault" labelText="I am checked by default" checked />
  <Checkbox name="disabled" labelText="I am disabled" disabled />
  <Checkbox name="uncheckedByDefault" labelText="I am checked and disabled" checked disabled />
</div>
```

Passing a function will execute with the current selection status of the Checkbox.

```
class CheckboxExample extends React.Component {
  constructor() {
    this.state = { 
      isAppleChecked: false,
      isBananaChecked: false,
      isStrawberryChecked: true
    }

    this.toggleCheckbox = this.toggleCheckbox.bind(this);
  }

  toggleCheckbox(key) {
    return event => {
      this.setState({ [key]: event.target.checked });
    }
  }

  render() {
    return (
      <div>
        <Checkbox 
          name="Apple"
          labelText="Apple"
          value="Apple"
          checked={this.state.isAppleChecked}
          onClick={this.toggleCheckbox('isAppleChecked')}
        />
        <Checkbox 
          name="Banana"
          labelText="Banana"
          value="Banana"
          checked={this.state.isBananaChecked}
          onClick={this.toggleCheckbox('isBananaChecked')}
        />
        <Checkbox 
          name="Strawberry"
          labelText="Strawberry"
          value="Strawberry"
          checked={this.state.isStrawberryChecked}
          onClick={this.toggleCheckbox('isStrawberryChecked')}
          disabled
        />
      </div>
    );
  }
}

<CheckboxExample />
```
### Validation States
```
class CheckboxExample extends React.Component {
  constructor() {
    this.state = { 
      isAppleChecked: false,
      isBananaChecked: false,
      isStrawberryChecked: false
    }

    this.toggleCheckbox = this.toggleCheckbox.bind(this);
  }

  toggleCheckbox(key) {
    return event => {
      this.setState({ [key]: event.target.checked });
    }
  }

  render() {
    return (
      <div>
        <Checkbox 
          name="Apple_Success"
          labelText="Apple"
          value="Apple"
          checked={this.state.isAppleChecked}
          onClick={this.toggleCheckbox('isAppleChecked')}
          validationState="success"
          additionalHelpContent="When validationState is set to success."
        />
        <Checkbox 
          name="Banana_Success"
          labelText="Banana"
          value="Banana"
          checked={this.state.isBananaChecked}
          onClick={this.toggleCheckbox('isBananaChecked')}
          validationState="warning"
          additionalHelpContent="When validationState is set to warning."
        />
        <Checkbox 
          name="Strawberry_danger"
          labelText="Strawberry"
          value="Strawberry"
          checked={this.state.isStrawberryChecked}
          onClick={this.toggleCheckbox('isStrawberryChecked')}
          validationState="danger"
          additionalHelpContent="When validationState is set to danger."
        />
      </div>
    );
  }
}

<CheckboxExample />
```
