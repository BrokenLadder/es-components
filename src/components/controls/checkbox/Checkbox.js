import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { noop } from 'lodash';

import Label from '../Label';
import { sizes, colors, inputColors } from '../../theme';

const CheckboxLabel = Label.extend`
  font-size: ${sizes.baseFontSize}px;
  font-weight: normal;
  text-transform: none;

  > .checkbox-fill {
    background-color: ${props => (props.isChecked ? colors.accent : colors.white)};
    border-color: ${props => (props.isChecked ? colors.accent : colors.grayDark)};

    &:after {
      border-color: ${props => (props.isChecked > colors.white: colors.grayDark)};
    }
  }

  &:hover > .checkbox-fill:after {
    border-color: ${props => (props.isChecked ? colors.white : colors.grayLight)};
  }

  &[disabled] > .checkbox-fill {
    background-color: ${props => (props.isChecked ? colors.gray : colors.white)};
    border-color: ${colors.gray};
    cursor: not-allowed;
    outline: 0;

    &:after {
      border-color: ${colors.white};
    }
  }
`;

const CheckboxInput = styled.input`
  clip: rect(0, 0, 0, 0);
  position: absolute;

  &:focus ~ .checkbox-fill {
    outline: 5px auto ${inputColors.inputDefaultFocus};
  }
`;

const CheckboxWrapper = styled.span`
  background: ${colors.white};
  border: 3px solid ${colors.grayDarker};
  border-radius: 4px;
  box-sizing: border-box;
  cursor: pointer;
  height: 25px;
  position: absolute;
  transition: all .25s linear;
  width: 25px;

  &:after {
    background: transparent;
    border: 3px solid ${colors.white};
    border-right: none;
    border-top: none;
    box-sizing: border-box;
    content: '';
    height: 9px;
    top: 3px;
    left: 2px;
    position: absolute;
    transform: rotate(-45deg);
    transition: border .25s linear;
    width: 15px;
  }
`;

const CheckboxText = styled.span`
  line-height: 1.4;
  margin-left: 30px;
`;

export default class Checkbox extends React.Component {
  static propTypes = {
    labelText: PropTypes.string.isRequired,
    initiallyChecked: PropTypes.bool,
    checkedValueChanged: PropTypes.func,
    disabled: PropTypes.bool
  };

  static defaultProps = {
    initiallyChecked: false,
    disabled: false,
    checkedValueChanged: noop
  };

  constructor(props) {
    super(props);

    this.handleOnClick = this.handleOnClick.bind(this);
  }

  state = {
    isChecked: this.props.initiallyChecked
  };

  componentWillReceiveProps({ initiallyChecked }) {
    this.setState(() => ({ isChecked: initiallyChecked }));
  }

  handleOnClick({ target }) {
    if (!this.props.disabled) {
      this.setState(previousState => ({ isChecked: !previousState.isChecked }));
      this.props.checkedValueChanged(target.checked);
    }
  }

  render() {
    const { labelText, disabled } = this.props;

    return (
      <CheckboxLabel disabled={disabled} isChecked={this.state.isChecked}>
        <CheckboxInput
          type="checkbox"
          checked={this.state.isChecked}
          onChange={this.handleOnClick}
          focusBorderColor={inputColors.inputDefaultFocus}
        />
        <CheckboxWrapper className="checkbox-fill" />
        <CheckboxText>{labelText}</CheckboxText>
      </CheckboxLabel>
    );
  }
}
