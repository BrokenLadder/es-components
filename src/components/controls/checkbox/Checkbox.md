```
<div>
  <Checkbox labelText="I am unchecked by default" />
  <Checkbox labelText="I am checked by default" initiallyChecked />
  <Checkbox labelText="I am disabled" disabled />
  <Checkbox labelText="I am checked and disabled" initiallyChecked disabled />
</div>
```

Passing a function will execute with the current selection status of the Checkbox.

```
function printIsChecked(isChecked) {
  console.log(`Is checked? ${isChecked}`);
}

<Checkbox labelText="Check the console when you check and uncheck me" checkedValueChanged={printIsChecked} />
```