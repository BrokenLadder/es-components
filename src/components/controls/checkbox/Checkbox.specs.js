/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import Checkbox from './Checkbox';

let instance;

function simulateLabelClick() {
  instance.find('input').simulate('change', { target: { checked: true } });
}

beforeEach(() => {
  instance = mount(<Checkbox labelText="test" />);
});

it('sets the state to the opposite of the previous state when label is clicked', () => {
  expect(instance.state('isChecked')).toBe(false);

  simulateLabelClick();

  expect(instance.state('isChecked')).toBe(true);
});

it('isChecked on the state is not updated if the isDisabled prop is true when clicked', () => {
  instance.setProps({ disabled: true });

  expect(instance.state('isChecked')).toBe(false);

  simulateLabelClick();

  expect(instance.state('isChecked')).toBe(false);
});

it('the checkedValueChanged prop gets executed with the value of the checkbox when clicked', () => {
  const checkedValueChanged = jest.fn();
  instance.setProps({ checkedValueChanged });

  simulateLabelClick();

  expect(checkedValueChanged.mock.calls.length).toBe(1);
  expect(checkedValueChanged.mock.calls[0][0]).toBe(true);
});

it('Checkbox renders as expected', () => {
  const tree = renderer.create(
    <Checkbox labelText="Render test" disabled initiallyChecked />
  );
  expect(tree).toMatchSnapshot();
});
