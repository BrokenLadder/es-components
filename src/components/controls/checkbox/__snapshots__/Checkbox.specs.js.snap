// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders as expected 1`] = `
<label
  className="sc-bwzfXH brRxqE"
  disabled={true}
>
  <input
    checked={true}
    className="sc-htpNat faSnhJ"
    disabled={true}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
    value={undefined}
  />
  <span
    className="checkbox-fill sc-bxivhb kjuRVg"
  />
  <span
    className="sc-ifAKCX gsuSvB"
  >
    Render test
  </span>
</label>
`;
